msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs

```diff
@@ -0,0 +1,202 @@
+
+            // Mock Handler for device code request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = CreateDeviceCodeResponseSuccessMessage()
+            });
+
+            // Mock Handler for devicecode->token exchange request
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                PostData = new Dictionary<string, string>()
+                {
+                    {OAuth2Parameter.ClientId, TestConstants.ClientId},
+                    {OAuth2Parameter.Scope, expectedScopes.AsSingleString()}
+                },
+                ResponseMessage = MockHelpers.CreateSuccessTokenResponseMessage()
+            });
+
+            DeviceCodeResult actualDeviceCodeResult = null;
+            DeviceCodeRequest request = new DeviceCodeRequest(parameters, result => { actualDeviceCodeResult = result; });
+            var task = request.RunAsync();
+            task.Wait();
+            var authenticationResult = task.Result;
+            Assert.IsNotNull(authenticationResult);
+            Assert.IsNotNull(actualDeviceCodeResult);
+
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+            Assert.AreEqual("BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA", actualDeviceCodeResult.DeviceCode);
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+
+            // todo: figure out how to unit test this one since it's calculated from current time based on the datetime offset retrieved in the response
+            // Assert.AreEqual(5, actualDeviceCodeResult.ExpiresOn); 
+
+            Assert.AreEqual("To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code B6SUYU5PL to authenticate.", actualDeviceCodeResult.Message);
+            Assert.AreEqual("B6SUYU5PL", actualDeviceCodeResult.UserCode);
+            Assert.AreEqual("https://microsoft.com/devicelogin", actualDeviceCodeResult.VerificationUrl);
+
+            Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
+        }
+
+        // todo: this is my test sample app that I built as a standalone console app.
+        // Need to figure out where we can check this in to do manual testing 
+        // of real device code flow login until we get automation in place on the
+        // MSAL side.
+        private static class __Program

```

Comment:
>__Program [](start = 29, length = 9)

seems like we typically would put this in our developer sample apps?

Reply:
yes, this is for the developer sample app.  I just didn't want to lose this snippet while I worked out getting the sample in place and whether there's a place for this in our internal dev/test apps instead of just an external sample.  since this makes testing the app for real actually possible.  would appreciate guidance here.

Reply:
Add it to the one of the dev sample apps. My recommendation would be to start with the desktop app.